pkg/image/prune/controller.go

```diff
@@ -0,0 +1,185 @@
+package prune
+
+import (
+	"fmt"
+	"strings"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+	gonum "github.com/gonum/graph"
+	"github.com/openshift/origin/pkg/api/graph"
+	"github.com/openshift/origin/pkg/client"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type PruningController struct {
+	images                 client.ImageInterface
+	streams                client.ImageStreamInterface
+	pods                   kclient.PodInterface
+	replicationControllers kclient.ReplicationControllerInterface
+	registryURL            string
+}
+
+func (c *PruningController) Run() error {
+	images, err := c.images.List(labels.Everything(), fields.Everything())
+	if err != nil {
+		return fmt.Errorf("error listing all images: %v", err)
+	}
+
+	streams, err := c.streams.List(labels.Everything(), fields.Everything())
+	if err != nil {
+		return fmt.Errorf("error listing all image streams: %v", err)
+	}
+
+	pods, err := c.pods.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("error listing all pods: %v", err)
+	}
+
+	rcs, err := c.replicationControllers.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("error listing all replication controllers: %v", err)
+	}
+
+	g := graph.New()
+	c.addImagesToGraph(g, images)
+	c.addImageStreamsToGraph(g, streams)
+	c.addPodsToGraph(g, pods)
+	c.addReplicationControllersToGraph(g, rcs)
+	c.prune(g)
+
+	return nil
+}
+
+const minimumImageAgeInMinutesToPrune = 60
+
+func (c *PruningController) addImagesToGraph(g graph.Graph, images *imageapi.ImageList) {
```

Comment:
Create 

type PruneImagesAlgorithm struct {
  // things needed for prune
}

func (PruneImagesAlgorithm) Run() {}

I'd like the traversal code be separate from the graph setup code separate from the actual "things to delete".  For instance, creating the image graph is one step NewPruneImagesAlgorithm(), you call Run(fn func(Image)) or something and return a count of the number of images affected.

----- Original Message -----
> > +		return fmt.Errorf("error listing all replication controllers: %v", err)
> > +	}
> > +
> > +	g := graph.New()
> > +	c.addImagesToGraph(g, images)
> > +	c.addImageStreamsToGraph(g, streams)
> > +	c.addPodsToGraph(g, pods)
> > +	c.addReplicationControllersToGraph(g, rcs)
> > +	c.prune(g)
> > +
> > +	return nil
> > +}
> > +
> > +const minimumImageAgeInMinutesToPrune = 60
> > +
> > +func (c *PruningController) addImagesToGraph(g graph.Graph, images
> > *imageapi.ImageList) {
> 
> I considered that, but I need c.registryURL. Do you want me to pass that in?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1841/files#r28803613